A52-xds-custom-lb-policies.md

```diff
@@ -0,0 +1,318 @@
+4. The registry entry for the endpoint picking policy notices that `TypeStruct` is being used, indicating a custom LB policy, and that policy is found to exist in the gRPC LB policy registry. The configuration is returned directly in its JSON form.
+5. The returned custom endpoint picking policy config is included in the `xds_wrr_locality` config by the WrrLocality conversion code which is returned to the xDS client.
+6. The xDS client calls the gRPC LB policy registry to parse the config converted by the xDS LB policy registry.
+7. The configuration parses, validating its correctness and the config is added to CDS update.
+
+#### 2. CDS load balancer builds the configuration for `xds_cluster_resolver`, including the configuration it received from xDS client. 
+
+```json
+{
+  "discovery_mechanisms": {
+    …
+  }
+  "xds_lb_policy": {
+    [
+      {
+        "xds_wrr_locality_experimental": {
+          "child_policy": [{
+            "myorg.MyCustomLeastRequestPolicy": {
+              "choiceCount": 2
+            },
+          }]
+        }
+      }
+    ]
+  }
+}
+```
+
+#### 3. The `xds_cluster_resolver` will use what it gets to generate the child configs for the priority LB policy.
+
+The priority config will contain the config for `xds_cluster_impl` that then has a child policy config for `xds_wrr_locality`. Assuming a single priority:
+
+```json
+{
+  "children": {
+    "priority-1": {
+      "config": [{
+        "xds_cluster_impl_experimental": {
+          "cluster": "ClusterName"
+          "child_policy": {
+            "xds_wrr_locality_experimental": {
+              "child_policy": [{
+                "myorg.MyCustomLeastRequestPolicy": {
+                  "choiceCount": 2
+                },
+              }]
+            }
+          }
+        }
+      }]
+      "ignore_reresolution_requests": false
+    }
+  }
+  "priorities": [ "priority-1" ]
+}
+```
+
+
+#### 4. `wrr_locality` combines the priority child config with locality weights to generate `weighted_target` configuration.
+
+Given the following locality weight attribute map:
+
+```json
+"Locality A": 1
+"Locality B": 2
+```
+
+The `weighted_target` configuration will be:
+
+```json
+{
+  "targets": {
+    "Locality A" : {
+      "weight": 1
+      "child_policy": [
+        {
+          "myorg.MyCustomLeastRequestPolicy" : {
+            "choiceCount" : 2
+          }
+        },
+      }
+    }
+    "Locality B" : {
+      "weight": 2
+      "child_policy": [
+        {
+          "myorg.MyCustomLeastRequestPolicy" : {
+            "choiceCount" : 2
+          }
+        },
+      }
+    }
+  }
+}
+```
+
+## Implementation
+Some initial refactoring to accommodate this desing has happened for Java. The Go library is yet to pick this up. In C-core the APIs that allow users to register custom LB policies are currently private. The plan is to make them public and once done this design can also be picked up for C-core.
```

Comment:
I don't think this needs to describe the current state of the implementation in each language.  If there are no implementation-specific details to describe, the "Implementation" section is not needed.

For C-core, I think this should say something like the following:

"""
In C-core, the LB policy API is not yet public, but it will be made public in the future when the EventEngine migration is far enough along to allow it.  In the interim, C-core will get the same changes to support custom xDS-based LB policies, so that when the C-core LB policy API later becomes public, this functionality will already be present.
"""

The other languages probably don't need to be mentioned at all.